// sasslint/lib/Linters/BorderZero.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _Linter2 = require('./Linter');

var _Linter3 = _interopRequireDefault(_Linter2);

var _lodashCollectionIncludes = require('lodash/collection/includes');

var _lodashCollectionIncludes2 = _interopRequireDefault(_lodashCollectionIncludes);

var BORDER_PROPERTIES = ['border', 'border-top', 'border-right', 'border-bottom', 'border-left'];

/**
 * @class BorderZero
 *
 * Prefer the terser `border: 0` over `border: none`.
 */

var BorderZero = (function (_Linter) {
    _inherits(BorderZero, _Linter);

    function BorderZero() {
        _classCallCheck(this, BorderZero);

        _get(Object.getPrototypeOf(BorderZero.prototype), 'constructor', this).apply(this, arguments);

        this.nodeTypes = ['Declaration'];
        this.defaults = {
            severity: 'warning',
            convention: '0'
        };
    }

    _createClass(BorderZero, [{
        key: 'run',
        value: function run(node) {
            var bannedFormat = this.options.convention === '0' ? 'none' : '0';

            if ((0, _lodashCollectionIncludes2['default'])(BORDER_PROPERTIES, node.property)) {
                if ((0, _lodashCollectionIncludes2['default'])(node.value, bannedFormat)) {
                    this.error(node, '\'border: ' + this.options.convention + '\' is preferred over \'border: ' + bannedFormat + '\'.');
                }
            }

            return null;
        }
    }]);

    return BorderZero;
})(_Linter3['default']);

exports['default'] = BorderZero;
module.exports = exports['default'];

/**
 * Types of nodes that this linter is interested in.
 * @see https://github.com/DFurnes/sasstree/tree/master/src/Nodes
 * @type {Array}
 */

/**
 * Default options.
 * @type {object}
 */
